src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from dataclasses import dataclass
+from pathlib import PurePath
+from typing import Tuple
+
+from pants.backend.python.target_types import PythonSources
+from pants.base.specs import AddressSpecs, AscendantAddresses
+from pants.core.util_rules.strip_source_roots import (
+    SourceRootStrippedSources,
+    StripSourcesFieldRequest,
+)
+from pants.engine.addresses import Address
+from pants.engine.collection import DeduplicatedCollection
+from pants.engine.rules import rule
+from pants.engine.selectors import Get, MultiGet
+from pants.engine.target import Targets
+from pants.source.source_root import AllSourceRoots
+
+
+@dataclass(frozen=True)
+class PythonModule:
+    module: str
+
+    @classmethod
+    def create_from_stripped_path(cls, path: PurePath) -> "PythonModule":
+        module_name_with_slashes = (
+            path.parent if path.name == "__init__.py" else path.with_suffix("")
+        )
+        return cls(module_name_with_slashes.as_posix().replace("/", "."))
+
+    def address_spec(self, *, source_root: str) -> AscendantAddresses:
+        """The spec for all candidate targets which could feasibly own the module.
+
+        This uses AscendantAddresses because targets can own files in subdirs (e.g. rglobs), but not
+        parents. We also use the package path, e.g. `helloworld/util/__init__.py`, rather than the
+        module path to ensure that we capture all possible targets. It is okay if this directory
+        does not actually exist.
+        """
+        module_name_with_slashes = self.module.replace(".", "/")
+        return AscendantAddresses(
+            directory=PurePath(source_root, module_name_with_slashes).as_posix()
+        )
```

Comment:
Two optimizations I see:
1. The `AscendantAddresses` object for each `import * source_root` will be effectively unique, so relatively little reuse can happen between these. It would probably be more efficient to list all targets below each source root with `DescendantAddresses`, (because it would not be module specific). Those could then be re-used to match owned sources.
2. The expansion of `AddressSpecs` is batched (the motivation was to be able to handle a case like "`--spec-file` containing a line per target in the repo"), so the effect of putting all of the `AscendantAddresses` for the imports of one target in one`AddressSpecs` object is that none of the per-import expansion of `AscendantAddresses` is reused _across_ targets. Instead, if each import was a separate `AddressSpecs` object, then would be memoized.

Reply:
Hm, how is this different than the original Toolchain implementation of having a mapping of every single target be pre-computed? I avoided that to make it be more lazy, so that we only evaluate folders we know may be in play - we know `pants.util.strutil` cannot live in `pants/backend/`.

Note that we do not know which source roots have the path at the time of calculating candidate_targets. We use `AscendantAddresses` on every single source root, every time.

Reply:
> Hm, how is this different than the original Toolchain implementation of having a mapping of every single target be pre-computed?

It might not be.

>  I avoided that to make it be more lazy, so that we only evaluate folders we know may be in play - we know pants.util.strutil cannot live in pants/backend/.

After having listed all targets in the source root, you can do the equivalent of `AscendantAddresses` by filtering to only find candidates "above" the module (based on their `spec_path`) before actually expanding globs.